R/linter_tags.R

```diff
@@ -60,47 +60,61 @@ available_linters <- function(packages = "lintr", tags = NULL, exclude_tags = "d
 
   csv_file <- system.file("lintr", "linters.csv", package = packages)
 
-  # Make sure we always return a valid data frame
-  # data.frame(...) and as.data.frame(...) don't handle zero-row list-columns properly, so we need to manually create
-  # the structure.
-  empty_linters <- data.frame(linter = character(), package = character(), stringsAsFactors = FALSE)
-  empty_linters$tags <- list()
-
   if (!file.exists(csv_file)) {
-    return(empty_linters)
+    return(empty_linters())
   }
   available <- utils::read.csv(csv_file, encoding = "UTF-8", as.is = TRUE)
 
-  # Check that the csv file contains two character columns, named 'linter' and 'tags'.
-  # Otherwise, fallback to an empty data frame.
-  if (!all(c("linter", "tags") %in% colnames(available))) {
-    warning(
-      "`linters.csv` must contain the columns 'linter' and 'tags'.\nPackage '",
-      packages, "' is missing ",
-      paste0("'", setdiff(c("linter", "tags"), colnames(available)), "'", collapse = " and "),
-      "."
-    )
-    return(empty_linters)
-  } else if (nrow(available) == 0L) {
-    # Circumvent empty list problem in data.frame(...)
-    return(empty_linters)
+  if (!validate_linter_db(available, packages)) {
+    return(empty_linters())
   }
 
-  res <- data.frame(
+  build_available_linters(available, packages, tags, exclude_tags)
+}
+
+build_available_linters <- function(available, packages, tags, exclude_tags) {
```

Comment:
`package` should be singular here, because the vectorization is already unrolled